File: example/py/test_error_callback_api.py
# /**
#  * test_error_callback_api.py
#  * @brief 错误/警告回调示例程序
#  * @attention
#  *   - 确保控制器网络通信正常
#  *   - 回调函数内不要做耗时操作或阻塞
#  *   - 默认控制器 IP 为 192.168.1.13，可根据实际环境修改
#  * @note 运行步骤
#  *       运行: python3 test_error_callback_api.py
#  */

import sys
sys.path.append("../../lib/linux/x86/python")
from tl_interface import *
import time


# ============================================================
# 回调函数（由用户实现，在收到错误/警告时被调用）
# ============================================================

def error_warning_callback(message_type: int, message: str, message_code: int):
    """
    处理接收到的错误和警告消息
    message_type: 消息类型
    message: 消息内容
    message_code: 消息代码
    """
    print(f"[回调] 类型={message_type}, 信息={message}, 代码={message_code}")


# ============================================================
# 工具函数
# ============================================================

def print_result(msg, ret):
    """打印操作结果"""
    if ret == SUCCESS:
        print("  [成功] " + msg)
    else:
        print("  [失败] " + msg + "，错误码: " + str(ret))


def print_separator(title):
    """打印分隔标题"""
    print("\n========== " + title + " ==========")


def enable_servo(socket_fd):
    """使能上电 —— 先清错，再检查伺服状态，报警则报错退出，最后上电"""
    print("--- 检查伺服状态 ---")

    # 1. 先清除控制器错误和报警，再下电复位确保状态同步
    ret = clear_error(socket_fd)
    time.sleep(0.3)
    set_servo_poweroff(socket_fd)
    time.sleep(0.3)

    # 2. 检查伺服状态
    state = -1
    ret, state = get_servo_state(socket_fd, state)
    if ret != SUCCESS:
        print(f"[ERROR] 获取伺服状态失败: {ret}")
        return
    print(f"  [信息] 当前伺服状态: {state} (0=停止 1=就绪 2=报警 3=运行)")

    if state == 2:
        print("[ERROR] 伺服处于报警状态，无法上电，请排查报警原因后重试")
        return

    if state == 3:
        print("[INFO] 机械臂已在使能上电状态")
        return

    # 3. 上电流程
    if state == 0:
        print("  [信息] 伺服处于停止状态，设置就绪...")
        ret = set_servo_state(socket_fd, 1)
        if ret != SUCCESS and ret != OPERATION_NOT_ALLOWED:
            print(f"[ERROR] 设置伺服就绪失败: {ret}")
            return
        print_result("set_servo_state(1)", ret)
        time.sleep(0.5)

    print("  [信息] 上电使能...")
    ret = set_servo_poweron(socket_fd)
    print_result("set_servo_poweron", ret)
    if ret != SUCCESS:
        print(f"[ERROR] 上电失败: {ret}")
        return
    time.sleep(2)

    # 4. 确认结果
    _, state = get_servo_state(socket_fd, state)
    if state == 3:
        print("  [成功] 机械臂使能上电成功 (servo_state = " + str(state) + ")")
    else:
        print("  [失败] 机械臂使能上电失败 (servo_state = " + str(state) + ")")


# ============================================================
# 主程序
# ============================================================
if __name__ == "__main__":
    # ---- 获取 SDK 版本 ----
    version = get_library_version()
    print("[信息] SDK 版本: " + version)

    # ---- 连接控制器 ----
    print_separator("连接控制器")
    sock = connect_robot("192.168.1.13", "6001")
    if sock < 0:
        print("[ERROR] 连接控制器失败")
        sys.exit(-1)
    print("[信息] 连接成功 (socket = " + str(sock) + ")")

    # ============================================================
    # 1. 注册错误/警告回调
    # ============================================================
    print_separator("注册回调")
    ret = set_receive_error_or_warnning_message_callback(sock, error_warning_callback)
    # 注意: SWIG 绑定存在已知问题，set_receive_error_or_warnning_message_callback
    # 的返回值被硬编码为 None（回调实际已被注册到 error_warning_callback_registry），
    # 此处 None 视为成功
    if ret is None or ret == SUCCESS:
        print("  [成功] set_receive_error_or_warnning_message_callback")
    else:
        print("  [失败] set_receive_error_or_warnning_message_callback，错误码: " + str(ret))
        disconnect_robot(sock)
        sys.exit(-1)
    print("  [信息] 回调已注册，后续收到错误/警告时会自动触发")

    # ============================================================
    # 2. 切换示教模式
    # ============================================================
    print_separator("切换示教模式")
    ret = set_current_mode(sock, 0)
    print_result("set_current_mode(0=示教)", ret)

    # ============================================================
    # 3. 伺服上电
    # ============================================================
    print_separator("伺服上电")
    enable_servo(sock)

    # ============================================================
    # 4. 发送不可达运动指令，触发错误回调
    # ============================================================
    print_separator("发送不可达运动指令")

    print("--- 发送一个超出工作空间的关节角度，观察回调触发 ---")
    cmd = MoveCmd()
    cmd.targetPosType = PosType_data
    cmd.coord = 0              # 关节坐标系
    cmd.velocity = 20          # 低速，安全
    cmd.acc = 10
    cmd.dec = 10
    cmd.pl = 0
    # 14 维向量: 前7本体 + 后7外部轴，J1=999° 远超限位
    pos = VectorDouble([0.0] * 14)
    pos[0] = 999
    cmd.targetPosValue = pos
    ret = robot_movej(sock, cmd)
    print_result("robot_movej (不可达关节角度)", ret)
    if ret != SUCCESS:
        print("  [信息] 运动失败，控制器的错误消息应已通过回调输出")
    else:
        print("  [信息] 运动指令已发送（意外成功，等待反馈...）")

    # 等待回调处理完成
    time.sleep(1)

    # ============================================================
    # 5. 断开连接
    # ============================================================
    print_separator("断开连接")
    disconnect_robot(sock)
    print("[信息] 已断开连接")

    print("\n[信息] 错误回调示例程序运行完毕")

File: example/cpp/test_error_callback_api.cpp
/**
 * test_error_callback_api.cpp
 * @brief 错误/警告回调示例程序
 * @attention
 *   - 确保控制器网络通信正常
 *   - 回调函数内不要做耗时操作或阻塞
 *   - 默认控制器 IP 为 192.168.1.13，可根据实际环境修改
 * @note 运行步骤
 *       编译: cd build && cmake .. && make
 *       运行: ./test_error_callback
 */

#include "cpp/interface/tl_api.h"
#include <iostream>
#include <string>
#include <thread>
#include <chrono>

// ============================================================
// 回调函数（由用户实现，在收到错误/警告时被调用）
// ============================================================

/// @brief 错误/警告消息回调
/// @param messageType 消息类型（0=警告 1=错误）
/// @param message     消息内容
/// @param messageCode 消息码
void on_error_or_warning(int messageType, const char* message, int messageCode)
{
    const char* type_str = (messageType == 0) ? "警告" : "错误";
    std::cout << "\n>>> [回调触发] " << type_str
              << " | code=" << messageCode
              << " | " << message << std::endl;
}

// ============================================================
// 工具函数
// ============================================================

void print_result(const std::string& msg, int ret)
{
    if (ret == Result::SUCCESS) {
        std::cout << "  [成功] " << msg << std::endl;
    } else {
        std::cout << "  [失败] " << msg << "，错误码: " << ret << std::endl;
    }
}

void print_separator(const std::string& title)
{
    std::cout << "\n========== " << title << " ==========" << std::endl;
}

// ============================================================
// 主函数
// ============================================================

int main()
{
    // ---- 获取 SDK 版本 ----
    {
        std::string version = get_library_version();
        std::cout << "[信息] SDK 版本: " << version << std::endl;
    }

    // ---- 连接控制器 ----
    print_separator("连接控制器");
    SOCKETFD sock = connect_robot("192.168.1.13", "6001");
    if (sock < 0) {
        std::cerr << "[ERROR] 连接控制器失败" << std::endl;
        return -1;
    }
    std::cout << "[信息] 连接成功 (socket = " << sock << ")" << std::endl;

    Result ret;

    // ============================================================
    // 1. 注册错误/警告回调
    // ============================================================
    print_separator("注册回调");
    ret = set_receive_error_or_warnning_message_callback(sock, on_error_or_warning);
    print_result("set_receive_error_or_warnning_message_callback", ret);
    if (ret != Result::SUCCESS) {
        std::cerr << "[ERROR] 注册回调失败" << std::endl;
        disconnect_robot(sock);
        return -1;
    }
    std::cout << "  [信息] 回调已注册，后续收到错误/警告时会自动触发" << std::endl;

    // ============================================================
    // 2. 逆解不可达点，触发错误回调
    // ============================================================
    // 通过 get_origin_coord_to_target_coord 尝试逆解一个超出工作空间的点，
    // 控制器会返回错误/警告，从而触发注册的回调函数
    print_separator("逆解不可达点");

    std::cout << "--- 逆解一个超出工作空间的点，观察回调触发 ---" << std::endl;
    {
        // 设置一个明显不可达的直角坐标点（X 方向超出机械臂最大臂展）
        std::vector<double> unreachable_pose = {99999, 0, 0, 0, 0, 0};
        std::vector<double> joint_result;
        bool convert_state = false;

        ret = get_origin_coord_to_target_coord(sock, 1, unreachable_pose, 0, joint_result, convert_state);
        print_result("逆解不可达点 (直角→关节)", ret);
        std::cout << "  输入直角坐标 [mm, rad]: X=99999, Y=0, Z=0" << std::endl;
        if (ret != Result::SUCCESS) {
            std::cout << "  [信息] 逆解失败，控制器的错误消息已通过回调输出" << std::endl;
        }
    }

    // 等待回调处理完成
    std::this_thread::sleep_for(std::chrono::seconds(2));

    // ============================================================
    // 3. 断开连接
    // ============================================================
    print_separator("断开连接");
    disconnect_robot(sock);
    std::cout << "[信息] 已断开连接" << std::endl;

    std::cout << "\n[信息] 错误回调示例程序运行完毕" << std::endl;
    return 0;
}
